src/Mod/PartDesign/App/FeatureHole.cpp

```diff
@@ -1387,64 +1399,119 @@ App::DocumentObjectExecReturn *Hole::execute(void)
         else
             return new App::DocumentObjectExecReturn("Hole: Could not revolve sketch!");
 
-#if 0
-        // Make thread
 
-        // This code is work in progress; making threas in OCC is not very easy, so
-        // this work is postponed until later
+        // Make thread
         if (ModelActualThread.getValue()) {
-            BRepBuilderAPI_MakeWire mkThreadWire;
-            double z = 0;
-            double d_min = Diameter.getValue() + ThreadCutOffInner.getValue();
-            double d_maj = Diameter.getValue() - ThreadCutOffInner.getValue();
-            int i = 0;
-
-            firstPoint = toPnt(xDir * d_min);
-
-            mkThreadWire.Add(BRepBuilderAPI_MakeEdge(gp_Pnt(0, 0, 0), firstPoint));
-            while (z < length) {
-                double z1 = i * ThreadPitch.getValue() + ThreadPitch.getValue() * 0.1;
-                double z2 = i * ThreadPitch.getValue() + ThreadPitch.getValue() * 0.45;
-                double z3 = i * ThreadPitch.getValue() + ThreadPitch.getValue() * 0.55;
-                double z4 = i * ThreadPitch.getValue() + ThreadPitch.getValue() * 0.9;
-
-                gp_Pnt p2 = toPnt(xDir * d_min - zDir * z1);
-                gp_Pnt p3 = toPnt(xDir * d_maj - zDir * z2);
-                gp_Pnt p4 = toPnt(xDir * d_maj - zDir * z3);
-                gp_Pnt p5 = toPnt(xDir * d_min - zDir * z4);
-
-                mkThreadWire.Add(BRepBuilderAPI_MakeEdge(firstPoint, p2));
-                mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p2, p3));
-                mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p3, p4));
-                mkThreadWire.Add(BRepBuilderAPI_MakeEdge(p4, p5));
-                firstPoint = p5;
-
-                ++i;
-                z += ThreadPitch.getValue();
-            }
-            mkThreadWire.Add(BRepBuilderAPI_MakeEdge(firstPoint, toPnt(-z * zDir)));
-            mkThreadWire.Add(BRepBuilderAPI_MakeEdge(toPnt(-z * zDir), gp_Pnt(0, 0, 0)));
+            const double pi = 3.14159265359;
+            bool leftHanded = (bool) ThreadDirection.getValue();
+
+            // Nomenclature and formulae according to Schaeffler Technical pocket guide (edition 2018)
+            // this is the same for all metric and UTS threads.
+            double D = threadDescription[ThreadType.getValue()][ThreadSize.getValue()].diameter;
+            double P = ThreadPitch.getValue();
+            double H1 = 0.54127 * P;
+            double H = 0.86603 * P;
+            double D1 = D - 2 * H1;
+            double alfa = 60.0 * pi/180;
+
+            // construct the cross section

```

Comment:
Also here, please code that others can understand it and also for yourself. I mean, for now it is basically clear, but do you know also next year why you used 8*tan and not 4*tan?

My proposal, comment the link to this drawing:
https://en.wikipedia.org/wiki/File:ISO_and_UTS_Thread_Dimensions.svg
 then comment every step you make and it should become clear what you do why.

Reply:
Good idea